File: src/main/kotlin/de/ljunker/queuedos/api/ApiModels.kt
package de.ljunker.queuedos.api

import de.ljunker.queuedos.domain.Priority
import de.ljunker.queuedos.domain.Role
import de.ljunker.queuedos.domain.SavedTicketFilterView
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class ApiError(
    val message: String,
    val code: String? = null,
    val currentVersion: Long? = null
)

@Serializable
data class LoginRequest(
    val email: String,
    val password: String
)

@Serializable
data class LoginResponse(
    val token: String,
    val user: UserResponse
)

@Serializable
data class AuthConfigResponse(
    val microsoftEnabled: Boolean
)

@Serializable
data class BootstrapResponse(
    val currentUser: UserResponse,
    val organizations: List<OrganizationResponse>,
    val users: List<UserResponse>,
    val projects: List<ProjectResponse>,
    val ticketTypes: List<TicketTypeResponse>,
    val workflows: List<WorkflowResponse>,
    val tickets: List<TicketResponse>,
    val deletedTickets: List<TicketResponse> = emptyList(),
    val savedTicketFilters: List<SavedTicketFilterResponse> = emptyList(),
    val activityHooks: List<ActivityHookResponse> = emptyList(),
    val priorities: List<Priority> = Priority.entries.toList()
)

@Serializable
data class TicketDetailResponse(
    val ticket: TicketResponse,
    val comments: List<TicketCommentResponse>,
    val revisions: TicketRevisionPageResponse,
    val legacyChanges: List<TicketChangeResponse>
)

@Serializable
data class TicketRevisionPageResponse(
    val revisions: List<TicketRevisionSummaryResponse>,
    val nextBeforeVersion: Long? = null
)

@Serializable
data class TicketRevisionSummaryResponse(
    val id: String,
    val ticketId: String,
    val version: Long,
    val actorId: String? = null,
    val action: String,
    val sourceVersion: Long? = null,
    val changes: List<TicketRevisionChangeResponse> = emptyList(),
    val createdAt: String,
    val restorable: Boolean
)

@Serializable
data class TicketRevisionDetailResponse(
    val revision: TicketRevisionSummaryResponse,
    val snapshot: TicketResponse
)

@Serializable
data class TicketRevisionChangeResponse(
    val field: String,
    val oldValue: JsonElement? = null,
    val newValue: JsonElement? = null
)

@Serializable
data class OrganizationResponse(
    val id: String,
    val name: String
)

@Serializable
data class UserResponse(
    val id: String,
    val organizationId: String,
    val email: String,
    val displayName: String,
    val role: Role,
    val active: Boolean
)

@Serializable
data class ProjectResponse(
    val id: String,
    val organizationId: String,
    val key: String,
    val name: String,
    val description: String = "",
    val color: String = "#2563eb",
    val nextTicketNumber: Int = 1,
    val archived: Boolean = false
)

@Serializable
data class TicketTypeResponse(
    val id: String,
    val organizationId: String,
    val projectId: String,
    val name: String,
    val description: String = "",
    val color: String = "#2563eb"
)

@Serializable
data class WorkflowStatusDto(
    val id: String,
    val name: String,
    val category: String = "TODO",
    val sortOrder: Int = 0
)

@Serializable
data class WorkflowTransitionDto(
    val id: String,
    val fromStatusId: String? = null,
    val toStatusId: String,
    val allowedRoles: List<Role> = listOf(Role.ADMIN, Role.MEMBER),
    val requiredFields: List<String> = emptyList(),
    val globalTransition: Boolean = false,
    val allowBackward: Boolean = true
)

@Serializable
data class WorkflowResponse(
    val id: String,
    val organizationId: String,
    val projectId: String,
    val statuses: List<WorkflowStatusDto>,
    val transitions: List<WorkflowTransitionDto>
)

@Serializable
data class TicketResponse(
    val id: String,
    val organizationId: String,
    val projectId: String,
    val number: Int,
    val key: String,
    val title: String,
    val description: String = "",
    val statusId: String,
    val typeId: String,
    val priority: Priority,
    val assigneeId: String? = null,
    val committedUserIds: List<String> = emptyList(),
    val labels: List<String> = emptyList(),
    val dueDate: String? = null,
    val estimate: Int? = null,
    val reporterId: String,
    val createdAt: String,
    val updatedAt: String,
    val deletedAt: String? = null,
    val deletedById: String? = null,
    val version: Long
)

@Serializable
data class TicketCommentResponse(
    val id: String,
    val organizationId: String,
    val ticketId: String,
    val authorId: String,
    val body: String,
    val createdAt: String
)

@Serializable
data class TicketChangeResponse(
    val id: String,
    val organizationId: String,
    val ticketId: String,
    val actorId: String,
    val field: String,
    val oldValue: String? = null,
    val newValue: String? = null,
    val createdAt: String
)

@Serializable
data class SavedTicketFilterCriteriaDto(
    val projectId: String? = null,
    val q: String = "",
    val statusId: String = "",
    val typeId: String = "",
    val priority: Priority? = null,
    val assigneeId: String = "",
    val label: String = "",
    val sort: String = "number"
)

@Serializable
data class SavedTicketFilterResponse(
    val id: String,
    val organizationId: String,
    val ownerId: String,
    val name: String,
    val view: SavedTicketFilterView,
    val projectId: String? = null,
    val filters: SavedTicketFilterCriteriaDto = SavedTicketFilterCriteriaDto()
)

@Serializable
data class ActivityHookResponse(
    val id: String,
    val organizationId: String,
    val eventType: de.ljunker.queuedos.domain.ActivityEventType,
    val webhookUrl: String,
    val messageTemplate: String,
    val active: Boolean
)

@Serializable
data class CreateProjectTicketTypeRequest(
    val name: String,
    val description: String = "",
    val color: String = "#2563eb"
)

@Serializable
data class CreateProjectStatusRequest(
    val name: String,
    val category: String
)

@Serializable
data class CreateProjectRequest(
    val key: String,
    val name: String,
    val description: String = "",
    val color: String = "#2563eb",
    val ticketTypes: List<CreateProjectTicketTypeRequest>? = null,
    val statuses: List<CreateProjectStatusRequest>? = null
)

@Serializable
data class UpdateProjectRequest(
    val key: String? = null,
    val name: String? = null,
    val description: String? = null,
    val archived: Boolean? = null,
    val color: String? = null
)

@Serializable
data class CreateUserRequest(
    val email: String,
    val displayName: String,
    val role: Role = Role.MEMBER,
    val password: String
)

@Serializable
data class UpdateUserRequest(
    val displayName: String? = null,
    val role: Role? = null,
    val active: Boolean? = null,
    val password: String? = null
)

@Serializable
data class CreateTicketTypeRequest(
    val projectId: String,
    val name: String,
    val description: String = "",
    val color: String = "#2563eb"
)

@Serializable
data class UpdateTicketTypeRequest(
    val name: String? = null,
    val description: String? = null,
    val color: String? = null
)

@Serializable
data class CreateTicketRequest(
    val projectId: String,
    val title: String,
    val description: String = "",
    val typeId: String,
    val priority: Priority = Priority.MEDIUM,
    val assigneeId: String? = null,
    val statusId: String? = null,
    val labels: List<String> = emptyList(),
    val dueDate: String? = null,
    val estimate: Int? = null
)

@Serializable
data class UpdateTicketRequest(
    val expectedVersion: Long,
    val title: String? = null,
    val description: String? = null,
    val typeId: String? = null,
    val priority: Priority? = null,
    val assigneeId: String? = null,
    val clearAssignee: Boolean = false,
    val labels: List<String>? = null,
    val dueDate: String? = null,
    val estimate: Int? = null,
    val clearDueDate: Boolean = false,
    val clearEstimate: Boolean = false,
    val statusId: String? = null
)

@Serializable
data class TransitionTicketRequest(
    val toStatusId: String,
    val expectedVersion: Long
)

@Serializable
data class CreateTicketCommentRequest(
    val body: String
)

@Serializable
data class SaveTicketCommitmentRequest(
    val committed: Boolean,
    val expectedVersion: Long
)

@Serializable
data class RestoreTicketRequest(val expectedVersion: Long)

@Serializable
data class SaveWorkflowRequest(
    val statuses: List<WorkflowStatusDto>,
    val transitions: List<WorkflowTransitionDto>
)

@Serializable
data class CreateSavedTicketFilterRequest(
    val name: String,
    val view: SavedTicketFilterView,
    val projectId: String? = null,
    val filters: SavedTicketFilterCriteriaDto = SavedTicketFilterCriteriaDto()
)

@Serializable
data class UpdateSavedTicketFilterRequest(
    val name: String? = null,
    val filters: SavedTicketFilterCriteriaDto? = null
)

@Serializable
data class BulkUpdateTicketsRequest(
    val tickets: List<VersionedTicketRefRequest>,
    val assigneeId: String? = null,
    val clearAssignee: Boolean = false,
    val priority: Priority? = null
)

@Serializable
data class VersionedTicketRefRequest(
    val id: String,
    val expectedVersion: Long
)

@Serializable
data class CreateActivityHookRequest(
    val eventType: de.ljunker.queuedos.domain.ActivityEventType,
    val webhookUrl: String,
    val messageTemplate: String,
    val active: Boolean = true
)

@Serializable
data class UpdateActivityHookRequest(
    val eventType: de.ljunker.queuedos.domain.ActivityEventType? = null,
    val webhookUrl: String? = null,
    val messageTemplate: String? = null,
    val active: Boolean? = null
)

File: frontend/src/app/core/api.models.ts
export type Role = 'ADMIN' | 'MEMBER';
export type Priority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface ApiError {
  message: string;
  code?: string | null;
  currentVersion?: number | null;
}

export interface Organization {
  id: string;
  name: string;
}

export interface PublicUser {
  id: string;
  organizationId: string;
  email: string;
  displayName: string;
  role: Role;
  active: boolean;
}

export interface Project {
  id: string;
  organizationId: string;
  key: string;
  name: string;
  description: string;
  color: string;
  nextTicketNumber: number;
  archived: boolean;
}

export interface TicketType {
  id: string;
  organizationId: string;
  projectId: string;
  name: string;
  description: string;
  color: string;
}

export interface WorkflowStatus {
  id: string;
  name: string;
  category: 'TODO' | 'IN_PROGRESS' | 'DONE' | string;
  sortOrder: number;
}

export interface WorkflowTransition {
  id: string;
  fromStatusId: string | null;
  toStatusId: string;
  allowedRoles: Role[];
  requiredFields: string[];
  globalTransition: boolean;
  allowBackward: boolean;
}

export interface Workflow {
  id: string;
  organizationId: string;
  projectId: string;
  statuses: WorkflowStatus[];
  transitions: WorkflowTransition[];
}

export interface Ticket {
  id: string;
  organizationId: string;
  projectId: string;
  number: number;
  key: string;
  title: string;
  description: string;
  statusId: string;
  typeId: string;
  priority: Priority;
  assigneeId: string | null;
  committedUserIds: string[];
  labels: string[];
  dueDate: string | null;
  estimate: number | null;
  reporterId: string;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
  deletedById: string | null;
  version: number;
}

export interface TicketComment {
  id: string;
  organizationId: string;
  ticketId: string;
  authorId: string;
  body: string;
  createdAt: string;
}

export interface TicketChange {
  id: string;
  organizationId: string;
  ticketId: string;
  actorId: string;
  field: string;
  oldValue: string | null;
  newValue: string | null;
  createdAt: string;
}

export type TicketRevisionAction =
  | 'BASELINE'
  | 'CREATED'
  | 'UPDATED'
  | 'MOVED'
  | 'COMMITMENT_CHANGED'
  | 'DELETED'
  | 'RESTORED'
  | 'REVISION_RESTORED';

export interface TicketRevisionChange {
  field: string;
  oldValue: unknown;
  newValue: unknown;
}

export interface TicketRevisionSummary {
  id: string;
  ticketId: string;
  version: number;
  actorId: string | null;
  action: TicketRevisionAction;
  sourceVersion: number | null;
  changes: TicketRevisionChange[];
  createdAt: string;
  restorable: boolean;
}

export interface TicketRevisionPage {
  revisions: TicketRevisionSummary[];
  nextBeforeVersion: number | null;
}

export interface TicketRevisionDetail {
  revision: TicketRevisionSummary;
  snapshot: Ticket;
}

export type SavedTicketFilterView = 'PROJECT_LIST' | 'MY_TICKETS';

export interface SavedTicketFilterCriteria {
  projectId: string | null;
  q: string;
  statusId: string;
  typeId: string;
  priority: Priority | null;
  assigneeId: string;
  label: string;
  sort: string;
}

export interface SavedTicketFilter {
  id: string;
  organizationId: string;
  ownerId: string;
  name: string;
  view: SavedTicketFilterView;
  projectId: string | null;
  filters: SavedTicketFilterCriteria;
}

export interface LoginRequest {
  email: string;
  password: string;
}

export interface LoginResponse {
  token: string;
  user: PublicUser;
}

export interface AuthConfigResponse {
  microsoftEnabled: boolean;
}

export type ActivityEventType =
    | 'TICKET_CREATED'
    | 'TICKET_UPDATED'
    | 'TICKET_MOVED'
    | 'COMMENT_ADDED'
    | 'COMMITMENT_CHANGED'
    | 'TICKET_DELETED'
    | 'TICKET_RESTORED';

export interface ActivityHook {
  id: string;
  organizationId: string;
  eventType: ActivityEventType;
  webhookUrl: string;
  messageTemplate: string;
  active: boolean;
}

export interface BootstrapResponse {
  currentUser: PublicUser;
  organizations: Organization[];
  users: PublicUser[];
  projects: Project[];
  ticketTypes: TicketType[];
  workflows: Workflow[];
  tickets: Ticket[];
  deletedTickets: Ticket[];
  savedTicketFilters: SavedTicketFilter[];
  activityHooks: ActivityHook[];
  priorities: Priority[];
}

export interface TicketDetailResponse {
  ticket: Ticket;
  comments: TicketComment[];
  revisions: TicketRevisionPage;
  legacyChanges: TicketChange[];
}

export interface CreateTicketCommentRequest {
  body: string;
}

export interface SaveTicketCommitmentRequest {
  committed: boolean;
  expectedVersion: number;
}

export interface CreateProjectRequest {
  key: string;
  name: string;
  description?: string;
  color?: string;
  ticketTypes?: CreateProjectTicketTypeRequest[];
  statuses?: CreateProjectStatusRequest[];
}

export interface CreateProjectTicketTypeRequest {
  name: string;
  description?: string;
  color?: string;
}

export interface CreateProjectStatusRequest {
  name: string;
  category: 'TODO' | 'IN_PROGRESS' | 'DONE';
}

export interface UpdateProjectRequest {
  key?: string;
  name?: string;
  description?: string;
  archived?: boolean;
  color?: string;
}

export interface CreateUserRequest {
  email: string;
  displayName: string;
  role: Role;
  password: string;
}

export interface UpdateUserRequest {
  displayName?: string;
  role?: Role;
  active?: boolean;
  password?: string;
}

export interface CreateTicketTypeRequest {
  projectId: string;
  name: string;
  description?: string;
  color?: string;
}

export interface UpdateTicketTypeRequest {
  name?: string;
  description?: string;
  color?: string;
}

export interface CreateTicketRequest {
  projectId: string;
  title: string;
  description: string;
  typeId: string;
  priority: Priority;
  assigneeId: string | null;
  statusId: string | null;
  labels: string[];
  dueDate: string | null;
  estimate: number | null;
}

export interface UpdateTicketRequest {
  expectedVersion: number;
  title?: string;
  description?: string;
  typeId?: string;
  priority?: Priority;
  assigneeId?: string | null;
  clearAssignee?: boolean;
  labels?: string[];
  dueDate?: string | null;
  estimate?: number | null;
  clearDueDate?: boolean;
  clearEstimate?: boolean;
  statusId?: string | null;
}

export interface TransitionTicketRequest {
  toStatusId: string;
  expectedVersion: number;
}

export interface BulkUpdateTicketsRequest {
  tickets: VersionedTicketRef[];
  assigneeId?: string | null;
  clearAssignee?: boolean;
  priority?: Priority | null;
}

export interface VersionedTicketRef {
  id: string;
  expectedVersion: number;
}

export interface SaveWorkflowRequest {
  statuses: WorkflowStatus[];
  transitions: WorkflowTransition[];
}

export interface CreateSavedTicketFilterRequest {
  name: string;
  view: SavedTicketFilterView;
  projectId: string | null;
  filters: SavedTicketFilterCriteria;
}

export interface UpdateSavedTicketFilterRequest {
  name?: string;
  filters?: SavedTicketFilterCriteria;
}

export interface CreateActivityHookRequest {
  eventType: ActivityEventType;
  webhookUrl: string;
  messageTemplate: string;
  active?: boolean;
}

export interface UpdateActivityHookRequest {
  eventType?: ActivityEventType;
  webhookUrl?: string;
  messageTemplate?: string;
  active?: boolean;
}
